File: engine/stat_stages.py
# Stat stage mechanics: application, multipliers, and move-to-effect mapping

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .pokemon import Pokemon


# ============================================================
# STAGE MECHANICS
# ============================================================

MIN_STAGE = -6
MAX_STAGE = 6


def get_stage_multiplier(stage: int) -> tuple[int, int]:
    """Return (numerator, denominator) for a stat stage.

    Gen 2: (2 + max(0, stage)) / (2 + max(0, -stage))
    """
    num = 2 + max(0, stage)
    den = 2 + max(0, -stage)
    return num, den


def apply_stat_change(pokemon: Pokemon, stat: str, stages: int) -> int:
    """Apply stat stage change, clamped to [-6, +6]. Returns actual change."""
    current = pokemon.stat_stages[stat]
    new = max(MIN_STAGE, min(MAX_STAGE, current + stages))
    actual = new - current
    pokemon.stat_stages[stat] = new
    return actual


# ============================================================
# MOVE STAT EFFECTS
# ============================================================

# (stat, stages, target) where target is "self" or "opponent"
StatEffect = tuple[str, int, str]

MOVE_STAT_EFFECTS: dict[str, list[StatEffect]] = {
    # ---- Self-boosting status moves ----
    "Swords Dance": [("attack", +2, "self")],
    "Growth": [("special_attack", +1, "self")],
    "Meditate": [("attack", +1, "self")],
    "Sharpen": [("attack", +1, "self")],
    "Agility": [("speed", +2, "self")],
    "Amnesia": [("special_defense", +2, "self")],
    "Barrier": [("defense", +2, "self")],
    "Acid Armor": [("defense", +2, "self")],
    "Double Team": [("evasion", +1, "self")],
    "Minimize": [("evasion", +1, "self")],
    "Harden": [("defense", +1, "self")],
    "Withdraw": [("defense", +1, "self")],
    "Defense Curl": [("defense", +1, "self")],
    # Curse: non-Ghost gets atk+1/def+1/speed-1 (Ghost curse is unimplemented)
    "Curse": [("attack", +1, "self"), ("defense", +1, "self"), ("speed", -1, "self")],
    "Belly Drum": [("attack", +6, "self")],

    # ---- Opponent-lowering status moves ----
    "Growl": [("attack", -1, "opponent")],
    "Leer": [("defense", -1, "opponent")],
    "Tail Whip": [("defense", -1, "opponent")],
    "String Shot": [("speed", -1, "opponent")],
    "Screech": [("defense", -2, "opponent")],
    "Charm": [("attack", -2, "opponent")],
    "Scary Face": [("speed", -2, "opponent")],
    "Sweet Scent": [("evasion", -1, "opponent")],
    "Sand Attack": [("accuracy", -1, "opponent")],
    "Smokescreen": [("accuracy", -1, "opponent")],
    "Flash": [("accuracy", -1, "opponent")],
    "Cotton Spore": [("speed", -2, "opponent")],
    "Kinesis": [("accuracy", -1, "opponent")],

    # ---- Secondary effects on damaging moves (rolled via stat_chance) ----
    "Acid": [("defense", -1, "opponent")],
    "Bubble": [("speed", -1, "opponent")],
    "Bubble Beam": [("speed", -1, "opponent")],
    "Aurora Beam": [("attack", -1, "opponent")],
    "Psychic": [("special_defense", -1, "opponent")],
    "Constrict": [("speed", -1, "opponent")],
    "Mud-Slap": [("accuracy", -1, "opponent")],
    "Octazooka": [("accuracy", -1, "opponent")],
    "Icy Wind": [("speed", -1, "opponent")],
    "Crunch": [("special_defense", -1, "opponent")],
    "Shadow Ball": [("special_defense", -1, "opponent")],
    "Rock Smash": [("defense", -1, "opponent")],
    "Iron Tail": [("defense", -1, "opponent")],
    # self-boosting secondaries
    "Steel Wing": [("defense", +1, "self")],
    "Metal Claw": [("attack", +1, "self")],
    "Ancient Power": [
        ("attack", +1, "self"),
        ("defense", +1, "self"),
        ("special_attack", +1, "self"),
        ("special_defense", +1, "self"),
        ("speed", +1, "self"),
    ],
}

File: crystal_engine/src/stat_stages.rs
// Stat stage mechanics and move-to-effect mapping

pub const MIN_STAGE: i8 = -6;
pub const MAX_STAGE: i8 = 6;

/// Gen 2 stage multiplier: (2 + max(0, stage)) / (2 + max(0, -stage))
pub fn get_stage_multiplier(stage: i8) -> (i32, i32) {
    let num = 2 + stage.max(0) as i32;
    let den = 2 + (-stage).max(0) as i32;
    (num, den)
}

/// Apply stat stage change, clamped to [-6, +6]. Returns actual change.
pub fn apply_stat_change(stages: &mut [i8; 7], stat_idx: usize, change: i8) -> i8 {
    let current = stages[stat_idx];
    let new = current.saturating_add(change).clamp(MIN_STAGE, MAX_STAGE);
    let actual = new - current;
    stages[stat_idx] = new;
    actual
}

// Stat indices into the stat_stages array
pub const STAT_ATK: usize = 0;
pub const STAT_DEF: usize = 1;
pub const STAT_SPA: usize = 2;
pub const STAT_SPD: usize = 3;
pub const STAT_SPE: usize = 4;
pub const STAT_ACC: usize = 5;
pub const STAT_EVA: usize = 6;

pub fn stat_idx_from_name(name: &str) -> Option<usize> {
    match name {
        "attack" => Some(STAT_ATK),
        "defense" => Some(STAT_DEF),
        "special_attack" => Some(STAT_SPA),
        "special_defense" => Some(STAT_SPD),
        "speed" => Some(STAT_SPE),
        "accuracy" => Some(STAT_ACC),
        "evasion" => Some(STAT_EVA),
        _ => None,
    }
}

pub fn stat_name_from_idx(idx: usize) -> &'static str {
    match idx {
        STAT_ATK => "attack",
        STAT_DEF => "defense",
        STAT_SPA => "special_attack",
        STAT_SPD => "special_defense",
        STAT_SPE => "speed",
        STAT_ACC => "accuracy",
        STAT_EVA => "evasion",
        _ => "unknown",
    }
}

// Stat effect: (stat_index, stages, targets_self)
pub type StatEffect = (usize, i8, bool);

/// Lookup move stat effects by move name. Returns None if no effects.
pub fn move_stat_effects(name: &str) -> Option<&'static [StatEffect]> {
    match name {
        // self-boosting status moves
        "Swords Dance" => Some(&[(STAT_ATK, 2, true)]),
        "Growth" => Some(&[(STAT_SPA, 1, true)]),
        "Meditate" => Some(&[(STAT_ATK, 1, true)]),
        "Sharpen" => Some(&[(STAT_ATK, 1, true)]),
        "Agility" => Some(&[(STAT_SPE, 2, true)]),
        "Amnesia" => Some(&[(STAT_SPD, 2, true)]),
        "Barrier" => Some(&[(STAT_DEF, 2, true)]),
        "Acid Armor" => Some(&[(STAT_DEF, 2, true)]),
        "Double Team" => Some(&[(STAT_EVA, 1, true)]),
        "Minimize" => Some(&[(STAT_EVA, 1, true)]),
        "Harden" => Some(&[(STAT_DEF, 1, true)]),
        "Withdraw" => Some(&[(STAT_DEF, 1, true)]),
        "Defense Curl" => Some(&[(STAT_DEF, 1, true)]),
        "Curse" => Some(&[(STAT_ATK, 1, true), (STAT_DEF, 1, true), (STAT_SPE, -1, true)]),
        "Belly Drum" => Some(&[(STAT_ATK, 6, true)]),
        // opponent-lowering status moves
        "Growl" => Some(&[(STAT_ATK, -1, false)]),
        "Leer" => Some(&[(STAT_DEF, -1, false)]),
        "Tail Whip" => Some(&[(STAT_DEF, -1, false)]),
        "String Shot" => Some(&[(STAT_SPE, -1, false)]),
        "Screech" => Some(&[(STAT_DEF, -2, false)]),
        "Charm" => Some(&[(STAT_ATK, -2, false)]),
        "Scary Face" => Some(&[(STAT_SPE, -2, false)]),
        "Sweet Scent" => Some(&[(STAT_EVA, -1, false)]),
        "Sand Attack" => Some(&[(STAT_ACC, -1, false)]),
        "Smokescreen" => Some(&[(STAT_ACC, -1, false)]),
        "Flash" => Some(&[(STAT_ACC, -1, false)]),
        "Cotton Spore" => Some(&[(STAT_SPE, -2, false)]),
        "Kinesis" => Some(&[(STAT_ACC, -1, false)]),
        // secondary effects on damaging moves
        "Acid" => Some(&[(STAT_DEF, -1, false)]),
        "Bubble" => Some(&[(STAT_SPE, -1, false)]),
        "Bubble Beam" => Some(&[(STAT_SPE, -1, false)]),
        "Aurora Beam" => Some(&[(STAT_ATK, -1, false)]),
        "Psychic" => Some(&[(STAT_SPD, -1, false)]),
        "Constrict" => Some(&[(STAT_SPE, -1, false)]),
        "Mud-Slap" => Some(&[(STAT_ACC, -1, false)]),
        "Octazooka" => Some(&[(STAT_ACC, -1, false)]),
        "Icy Wind" => Some(&[(STAT_SPE, -1, false)]),
        "Crunch" => Some(&[(STAT_SPD, -1, false)]),
        "Shadow Ball" => Some(&[(STAT_SPD, -1, false)]),
        "Rock Smash" => Some(&[(STAT_DEF, -1, false)]),
        "Iron Tail" => Some(&[(STAT_DEF, -1, false)]),
        "Steel Wing" => Some(&[(STAT_DEF, 1, true)]),
        "Metal Claw" => Some(&[(STAT_ATK, 1, true)]),
        "Ancient Power" => Some(&[
            (STAT_ATK, 1, true),
            (STAT_DEF, 1, true),
            (STAT_SPA, 1, true),
            (STAT_SPD, 1, true),
            (STAT_SPE, 1, true),
        ]),
        _ => None,
    }
}
